design/json-substitution-action-design.md

```diff
@@ -0,0 +1,122 @@
+- She can use this feature to substitute the DES in all StorageClass YAMLs with the new DES without having to create a fresh storageclass, or understanding the name of the storageclass.
+
+### Scenario 2
+- Bob has multi zone cluster where nodes are spread across zones. 
+- Bob has pinned certain pods to a particular zone using nodeSelector/ nodeaffinity on the pod spec.
+- In case of zone outage of the cloudprovider, Bob wishes to restore the workload to a different namespace in the same cluster, but change the zone pinning of the workload.
+- Bob can use this feature to substitute the nodeSelector/ nodeaffinity in the pod spec with the new zone pinning to quickly failover the workload to a different zone's nodes.
+
+## Detailed Design
+- The user is expected to create a configmap with the desired JSON Substitutions. Then the reference of the configmap will be provided in the RestoreSpec.
+- The core restore workflow before creating/updating a particular resource in the cluster will be checked against the filters provided and respective substitutions will be applied on it.
+
+Example of Reference to configmap in RestoreSpec
+```yaml
+apiVersion: velero.io/v1
+kind: Restore
+metadata:
+  name: restore-1
+spec:
+    jsonSubstitution:
+      refType: Configmap
+      ref: backup01
+```
+
+### ConfigMap Structure
+- User first needs to provide details on which resources the JSON Substitutions need to be applied. 
+    - For this the user will provide 4 inputs - Nameespaces(for NS Scoped resources), Group, Kind and Name Regex(optional).
+    - If the user does not provide the Name, the JSON Substitutions will be applied to all the resources of the given Group and Kind under the given namespaces.
+
+- Further the use will specify the JSON Substitutions using the structure of kubectl's "JSON Path" based inputs.
+    - Reference: [https://kubernetes.io/docs/reference/kubectl/jsonpath/](https://kubernetes.io/docs/reference/kubectl/jsonpath/)
+    - ![image](https://user-images.githubusercontent.com/36476228/219611516-7fa9527d-c592-4471-8e3e-39ad5e7606db.png)
+    - This filter can be used through golang Package: "k8s.io/client-go/util/jsonpath"
+
+- Last the user will provide the current value regex, basically at the path specified above, we check if the provided regex matches the value at that path. If it does, we substitute the value with the new value provided by the user.
+
+- Sample ConfigMap
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+    name: json-substitution-configmap
+    namespace: velero
+data:
+    json-substitution-configmap.yaml: |
+        resources:
+        - namespaces:
+            - default
+        group: storage.k8s.io
+        kind: StorageClass
+        nameRegex: "disk*"
+        jsonSubstitutions:
+        - path: {..parameters.diskEncryptionSetID}
+            currentValueRegex: ".*"
+            newValue: "new-des"
+        - namespaces:
+            - default
+        group: apps
```

Comment:
In a corner case, if two rules happen to modify the same field of same object, the latter will overwrite the earlier right?

Could you also clarify that in the design?

Reply:
Ack will do.